File: leanback/src/main/java/de/nicidienase/chaosflix/leanback/fragments/BrowseErrorFragment.java
package de.nicidienase.chaosflix.leanback.fragments;

import android.content.res.Resources;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v17.leanback.app.ErrorSupportFragment;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ProgressBar;

import de.nicidienase.chaosflix.leanback.R;

public class BrowseErrorFragment extends ErrorSupportFragment {

	private static final boolean TRANSLUCENT = true;
	public static final String FRAGMENT = "fragmentId";
	private SpinnerFragment spinnerFragment;

	public static BrowseErrorFragment showErrorFragment(FragmentManager manager, int fragmentId) {
		BrowseErrorFragment errorFragment = new BrowseErrorFragment();
		Bundle args = new Bundle();
		args.putInt(BrowseErrorFragment.FRAGMENT, fragmentId);
		errorFragment.setArguments(args);
		manager.beginTransaction().replace(fragmentId, errorFragment)
				.addToBackStack(null).commit();
		return errorFragment;
	}

	@Override
	public void onCreate(@Nullable Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		int fragmentId = getArguments().getInt(FRAGMENT);
		spinnerFragment = new SpinnerFragment();
		getFragmentManager().beginTransaction().add(fragmentId, spinnerFragment).commit();
	}

	public void setErrorContent(int resourceId) {
		setErrorContent(getResources().getString(resourceId));
	}

	public void setErrorContent(String message) {
		setImageDrawable(getResources().getDrawable(R.drawable.lb_ic_sad_cloud, null));
		setMessage(message);
		setDefaultBackground(TRANSLUCENT);

		setButtonText(getResources().getString(R.string.dismiss_error));
		setButtonClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View arg0) {
				dismiss();
			}
		});
	}

	public void dismiss() {
		FragmentManager fragmentManager = getFragmentManager();
		if (fragmentManager != null) {
			fragmentManager
					.beginTransaction()
					.remove(BrowseErrorFragment.this)
					.remove(spinnerFragment)
					.commit();
			fragmentManager.popBackStack();
		}
	}


	public static class SpinnerFragment extends Fragment {
		@Override
		public View onCreateView(LayoutInflater inflater, ViewGroup container,
								 Bundle savedInstanceState) {
			ProgressBar progressBar = new ProgressBar(container.getContext());
			if (container instanceof FrameLayout) {
				Resources res = getResources();
				int width = res.getDimensionPixelSize(R.dimen.spinner_width);
				int height = res.getDimensionPixelSize(R.dimen.spinner_height);
				FrameLayout.LayoutParams layoutParams =
						new FrameLayout.LayoutParams(width, height, Gravity.CENTER);
				progressBar.setLayoutParams(layoutParams);
			}
			return progressBar;
		}
	}
}

File: leanback/src/main/java/de/nicidienase/chaosflix/leanback/BrowseErrorFragment.kt
package de.nicidienase.chaosflix.leanback

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.leanback.app.ErrorSupportFragment

class BrowseErrorFragment : ErrorSupportFragment() {
    private var spinnerFragment: SpinnerFragment? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val fragmentId = arguments?.getInt(FRAGMENT) ?: throw IllegalStateException("missing fragmentID")
        spinnerFragment = SpinnerFragment().apply {
            try {
                parentFragmentManager.beginTransaction().add(fragmentId, this).commit()
            } catch (ex: IllegalStateException) {
            }
        }
    }

    fun setErrorContent(resourceId: Int) {
        setErrorContent(resources.getString(resourceId))
    }

    fun setErrorContent(message: String, parentFragmentManager: androidx.fragment.app.FragmentManager? = activity?.supportFragmentManager) {
        try {
            if (!isDetached) {
                spinnerFragment?.let {
                    parentFragmentManager?.beginTransaction()?.remove(it)?.commit()
                }
                imageDrawable = resources.getDrawable(R.drawable.lb_ic_sad_cloud, null)
                setMessage(message)
                setDefaultBackground(TRANSLUCENT)
                buttonText = resources.getString(R.string.dismiss_error)

                if (parentFragmentManager != null) {
                    setButtonClickListener { _ -> dismiss(parentFragmentManager) }
                } else {
                    setButtonClickListener { _ -> dismiss() }
                }
            }
        } catch (ex: IllegalStateException) {
            Log.e(TAG, "could not show error fragment")
        }
    }

    override fun onPause() {
        super.onPause()
        spinnerFragment?.let {
            parentFragmentManager?.beginTransaction()?.remove(it)?.commit()
        } ?: Log.e(TAG, "Could not remove spinnerFragment")
    }

    fun dismiss(parentFragmentManager: androidx.fragment.app.FragmentManager? = activity?.supportFragmentManager) {
        if (parentFragmentManager != null) {
            with(parentFragmentManager.beginTransaction()) {
                spinnerFragment?.let {
                    remove(it)
                }
                remove(this@BrowseErrorFragment)
                commit()
                parentFragmentManager.popBackStack()
            }
        } else {
            Log.e(TAG, "Cannot dismiss, parentFragmentManager is null")
        }
    }

    class SpinnerFragment : androidx.fragment.app.Fragment() {
        override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
            return inflater.inflate(R.layout.loading_fragment, container, false)
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        spinnerFragment = null
    }

    companion object {

        private val TRANSLUCENT = true
        val FRAGMENT = "fragmentId"
        val TAG = BrowseErrorFragment::class.java.simpleName

        fun showErrorFragment(manager: androidx.fragment.app.FragmentManager, fragmentId: Int, addToBackstack: Boolean = false): BrowseErrorFragment {
            val errorFragment = BrowseErrorFragment()
            val args = Bundle()
            args.putInt(BrowseErrorFragment.FRAGMENT, fragmentId)
            errorFragment.arguments = args
            if (addToBackstack) {
                manager.beginTransaction().add(fragmentId, errorFragment).addToBackStack(null).commit()
            } else {
                manager.beginTransaction().add(fragmentId, errorFragment).commit()
            }
            return errorFragment
        }
    }
}
